File: backend/app/services/feedback_service.py
"""
backend/app/services/feedback_service.py
------------------------------------------
Business logic for investigator feedback on similarity results.

Closes the human-in-the-loop cycle:
    Investigator flags wrong result → feedback saved →
    AI engineer exports incorrect cases → retrain_from_feedback.py
    improves the model → better future predictions.

Works for both fingerprint and toolmark results — the feedback record
links to a SimilarityResult which already carries the evidence_type
via its linked ForensicImage.
"""

from typing import Optional

from fastapi import HTTPException, status
from sqlalchemy import select, desc, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.feedback          import Feedback
from app.models.similarity_result import SimilarityResult
from app.models.user              import User
from app.schemas.feedback_schema  import (
    FeedbackCreate,
    FeedbackResponse,
    FeedbackListResponse,
)
from app.services.log_service import create_log


# ---------------------------------------------------------------------------
async def submit_feedback(
    payload:    FeedbackCreate,
    user:       User,
    db:         AsyncSession,
    ip_address: Optional[str] = None,
) -> FeedbackResponse:
    """
    Submit investigator feedback on a similarity result.

    Validates:
        - SimilarityResult exists and is accessible by the user
        - User hasn't already submitted feedback for this result

    Raises:
        HTTP 404 — result not found
        HTTP 403 — result belongs to another user (non-admin)
        HTTP 409 — feedback already submitted for this result by this user
    """
    # Validate result exists and is accessible
    row = await db.execute(
        select(SimilarityResult).where(SimilarityResult.id == payload.result_id)
    )
    result = row.scalar_one_or_none()

    if result is None:
        raise HTTPException(
            status_code = status.HTTP_404_NOT_FOUND,
            detail      = f"Similarity result {payload.result_id} not found.",
        )
    if result.requested_by != user.id and user.role != "admin":
        raise HTTPException(
            status_code = status.HTTP_403_FORBIDDEN,
            detail      = "Access denied.",
        )

    # Prevent duplicate feedback from the same user on the same result
    existing = await db.execute(
        select(Feedback).where(
            Feedback.result_id == payload.result_id,
            Feedback.user_id   == user.id,
        )
    )
    if existing.scalar_one_or_none():
        raise HTTPException(
            status_code = status.HTTP_409_CONFLICT,
            detail      = (
                f"You have already submitted feedback for result "
                f"{payload.result_id}."
            ),
        )

    feedback = Feedback(
        user_id    = user.id,
        result_id  = payload.result_id,
        is_correct = payload.is_correct,
        comment    = payload.comment,
    )
    db.add(feedback)
    await db.commit()
    await db.refresh(feedback)

    # Audit log
    await create_log(
        db          = db,
        action_type = "feedback_submitted",
        user_id     = user.id,
        details     = {
            "feedback_id": feedback.id,
            "result_id":   payload.result_id,
            "is_correct":  payload.is_correct,
        },
        ip_address  = ip_address,
    )

    return FeedbackResponse.model_validate(feedback)


# ---------------------------------------------------------------------------
async def list_feedback(
    db:         AsyncSession,
    page:       int           = 1,
    limit:      int           = 50,
    is_correct: Optional[bool] = None,
) -> FeedbackListResponse:
    """
    List all feedback records — admin and AI engineer only.

    Supports filtering by is_correct:
        is_correct=False → only incorrect predictions (for retraining)
        is_correct=True  → confirmed correct predictions
        is_correct=None  → all feedback

    Returns summary counts for the admin dashboard widget.
    """
    limit = min(limit, 100)
    query = select(Feedback).order_by(desc(Feedback.created_at))

    if is_correct is not None:
        query = query.where(Feedback.is_correct == is_correct)

    count_result = await db.execute(query.with_only_columns(Feedback.id))
    total = len(count_result.all())

    offset   = (page - 1) * limit
    rows     = await db.execute(query.offset(offset).limit(limit))
    feedback = rows.scalars().all()

    # Summary counts for dashboard widget
    correct_count = await db.execute(
        select(func.count(Feedback.id)).where(Feedback.is_correct == True)
    )
    incorrect_count = await db.execute(
        select(func.count(Feedback.id)).where(Feedback.is_correct == False)
    )

    return FeedbackListResponse(
        total            = total,
        page             = page,
        limit            = limit,
        feedback         = [FeedbackResponse.model_validate(f) for f in feedback],
        total_correct    = correct_count.scalar(),
        total_incorrect  = incorrect_count.scalar(),
    )


# ---------------------------------------------------------------------------
async def get_incorrect_feedback(
    db: AsyncSession,
) -> list[dict]:
    """
    Export all incorrect feedback cases for retraining.

    Called by ai_engine/training/retrain_from_feedback.py to identify
    hard examples — image pairs the model got wrong according to
    investigators.

    Returns a list of dicts with:
        result_id, image_id_1, image_id_2,
        similarity_percentage, match_status, comment
    """
    rows = await db.execute(
        select(Feedback, SimilarityResult)
        .join(SimilarityResult, Feedback.result_id == SimilarityResult.id)
        .where(Feedback.is_correct == False)
        .order_by(desc(Feedback.created_at))
    )

    results = []
    for feedback, result in rows.all():
        results.append({
            "feedback_id":           feedback.id,
            "result_id":             result.id,
            "image_id_1":            result.image_id_1,
            "image_id_2":            result.image_id_2,
            "similarity_percentage": result.similarity_percentage,
            "match_status":          result.match_status,
            "investigator_comment":  feedback.comment,
        })

    return results

File: frontend/src/services/feedbackService.ts
/**
 * src/services/feedbackService.ts
 * ─────────────────────────────────
 * Investigator feedback on similarity results.
 *
 * Backend endpoints consumed
 * ───────────────────────────
 *   POST /api/v1/feedback          body: FeedbackCreate  → FeedbackResponse
 *   GET  /api/v1/feedback          query params          → FeedbackListResponse
 *                                  (admin / ai_engineer only)
 *   GET  /api/v1/feedback/{id}                           → FeedbackResponse
 *                                  (admin / ai_engineer only)
 *   GET  /api/v1/feedback/export                         → dict[]
 *                                  (admin / ai_engineer only — for retraining)
 *
 * Human-in-the-loop cycle
 * ────────────────────────
 *   Investigator flags wrong result  →  feedback stored in DB
 *   AI engineer exports incorrect cases via /feedback/export
 *   retrain_from_feedback.py builds a hard-example dataset
 *   Model is retrained → better future predictions
 *
 * Access control
 * ───────────────
 *   submit()  — any authenticated active user (analyst, admin, ai_engineer)
 *   list()    — admin and ai_engineer only (HTTP 403 for analysts)
 *   get()     — admin and ai_engineer only
 *
 * One submission per result per user
 * ────────────────────────────────────
 *   The backend rejects duplicate submissions with HTTP 409 CONFLICT.
 *   The frontend should disable the feedback buttons once submitted.
 */

import api from "./api";

// ── Types ────────────────────────────────────────────────────────────────────

/**
 * Matches FeedbackResponse in backend/app/schemas/feedback_schema.py
 *
 * is_correct — true: investigator confirms the prediction was right
 *              false: investigator flags it as wrong (used for retraining)
 * comment    — optional explanation, especially useful when is_correct=false
 */
export interface FeedbackResponse {
  id:         number;
  result_id:  number;
  user_id:    number;
  is_correct: boolean;
  comment:    string | null;
  created_at: string;
}

/**
 * Matches FeedbackListResponse
 *
 * total_correct   — count of feedbacks where is_correct=true
 * total_incorrect — count of feedbacks where is_correct=false
 * Both are optional (null if the backend omits them).
 */
export interface FeedbackListResponse {
  total:           number;
  page:            number;
  limit:           number;
  feedback:        FeedbackResponse[];
  total_correct:   number | null;
  total_incorrect: number | null;
}

/**
 * Shape of records returned by GET /feedback/export
 * Used by the AI team to build a retraining dataset.
 */
export interface FeedbackExportRecord {
  feedback_id:           number;
  result_id:             number;
  image_id_1:            number;
  image_id_2:            number;
  similarity_percentage: number;
  match_status:          string;
  investigator_comment:  string | null;
}

// ── Service ──────────────────────────────────────────────────────────────────

export const feedbackService = {

  /**
   * POST /api/v1/feedback
   * Submits investigator feedback on a similarity result.
   *
   * @param resultId   — ID of the SimilarityResult being reviewed
   * @param isCorrect  — true: model was right | false: model was wrong
   * @param comment    — optional explanation (strongly recommended when
   *                     isCorrect=false to aid the retraining team)
   *
   * @throws AxiosError 404 if result not found or not accessible
   * @throws AxiosError 409 if feedback already submitted for this result
   */
  async submit(
    resultId:   number,
    isCorrect:  boolean,
    comment?:   string,
  ): Promise<FeedbackResponse> {
    const { data } = await api.post<FeedbackResponse>("/feedback", {
      result_id:  resultId,
      is_correct: isCorrect,
      comment:    comment ?? null,
    });
    return data;
  },

  /**
   * GET /api/v1/feedback
   * Returns paginated feedback records.
   * Requires admin or ai_engineer role (HTTP 403 otherwise).
   *
   * @param params.isCorrect — filter: true=correct, false=incorrect, undefined=all
   * @param params.page      — 1-based page number (default 1)
   * @param params.limit     — records per page (default 50)
   */
  async list(params?: {
    isCorrect?: boolean;
    page?:      number;
    limit?:     number;
  }): Promise<FeedbackListResponse> {
    // Map camelCase to the snake_case query param the backend expects
    const { data } = await api.get<FeedbackListResponse>("/feedback", {
      params: {
        is_correct: params?.isCorrect,
        page:       params?.page,
        limit:      params?.limit,
      },
    });
    return data;
  },

  /**
   * GET /api/v1/feedback/{id}
   * Returns a single feedback record.
   * Requires admin or ai_engineer role.
   */
  async get(feedbackId: number): Promise<FeedbackResponse> {
    const { data } = await api.get<FeedbackResponse>(
      `/feedback/${feedbackId}`,
    );
    return data;
  },

  /**
   * GET /api/v1/feedback/export
   * Exports all incorrect feedback cases as structured data for retraining.
   * Requires admin or ai_engineer role.
   *
   * The returned records are consumed by
   * ai_engine/training/retrain_from_feedback.py.
   */
  async exportIncorrect(): Promise<FeedbackExportRecord[]> {
    const { data } = await api.get<FeedbackExportRecord[]>(
      "/feedback/export",
    );
    return data;
  },
};